presto-manager-agent/src/main/java/com/teradata/prestomanager/agent/InstallationManager.java

```diff
@@ -17,78 +17,147 @@
+            }
+            URL url = new URL(packageUrl);
+            PrestoAsyncCommand(() -> installAsync(url, checkDependencies));
+        }
+        catch (PrestoManagerException e) {
+            LOGGER.error(e, "Failed to ascertain whether presto is already installed.");
+            return javax.ws.rs.core.Response.status(INTERNAL_SERVER_ERROR).entity("Failed to ascertain whether presto is already installed.").build();
+        }
+        catch (MalformedURLException e) {
+            LOGGER.error(e, "Invalid url: %s", packageUrl);
+            return javax.ws.rs.core.Response.status(BAD_REQUEST).entity("Invalid url").build();
+        }
+        return javax.ws.rs.core.Response.status(ACCEPTED).entity("Presto is being installed.\r\n" +
+                "To verify that installation succeeded, check back later using the status API.").build();
     }
 
-    public Response startUsingRpm()
+    public Response uninstall(boolean checkDependencies, boolean forceUninstall)
     {
         try {
             if (!isInstalled()) {
                 LOGGER.error("Presto is not installed");
-                return Response.status(NOT_FOUND).entity("Presto is not installed").build();
+                return Response.status(BAD_REQUEST).entity("Presto is not installed").build();
+            }
+            if (isRunning()) {
+                if (!forceUninstall) {
+                    LOGGER.error("Presto is running. Stop Presto before beginning upgrade.");
+                    return Response.status(CONFLICT).entity("Presto is running. Stop Presto before beginning upgrade.").build();
+                }
+                LOGGER.warn("Presto is running; Presto will be forcibly stopped before attempting to uninstall.");
+                stop(TERMINATE);
             }
         }
         catch (PrestoManagerException e) {
-            LOGGER.error(e.getCause(), e.getMessage());
-            return Response.status(INTERNAL_SERVER_ERROR).entity(e.getMessage()).build();
+            LOGGER.error(e, "Failed to ascertain whether presto is already installed or running.");
+            return javax.ws.rs.core.Response.status(INTERNAL_SERVER_ERROR).entity("Failed to ascertain whether presto is already installed or running.").build();
         }
-        new Thread(() -> {
-            try {
-                int startPresto = executeCommand(SUBPROCESS_TIMEOUT, "service", "presto", "start");
-                if (startPresto != 0) {
-                    LOGGER.error("Failed to start presto");
+        PrestoAsyncCommand(() -> uninstallAsync(checkDependencies));
+        return Response.status(ACCEPTED).entity("Presto is being uninstalled.\r\n" +
+                "To verify that uninstallation succeeded, check back later using the status API.").build();
+    }
+
+    public Response upgrade(String packageUrl, boolean checkDependencies, boolean preserveConfig, boolean forceUpgrade)
```

Comment:
For upgrade presto not installed is not a problem. `rpm -U` doesn't throw any error if its already installed.